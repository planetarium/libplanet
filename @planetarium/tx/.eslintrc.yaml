ignorePatterns:
- dist/
env:
  browser: true
  es2021: true
extends:
- eslint:recommended
- plugin:@typescript-eslint/recommended
- prettier
overrides: []
parser: '@typescript-eslint/parser'
parserOptions:
  ecmaVersion: latest
  sourceType: module
  project: ./tsconfig.json
plugins:
- '@typescript-eslint'
rules:
  max-len: [error, { code: 80 }]
  indent: [error, 2]
  linebreak-style: [error, unix]
  quotes: [error, double]
  semi: [error, always]
  consistent-return: [error, { treatUndefinedAsUnspecified: true }]
  '@typescript-eslint/consistent-type-exports': error
