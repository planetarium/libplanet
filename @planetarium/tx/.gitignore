coverage/
dist/
node_modules/
package-lock.json
planetarium-tx-*.tgz
