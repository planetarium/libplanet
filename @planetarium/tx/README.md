`@planetarium/tx`
=================

[![npm][npm-badge]][npm]

This npm package provides functions to build transactions equivalent to
[Libplanet]:

 -  Creating an unsigned transaction with a system action
 - [ ] (TODO) Creating an unsigned transaction with custom actions
 - [ ] Signing a transaction

[npm]: https://www.npmjs.com/package/@planetarium/tx
[npm-badge]: https://img.shields.io/npm/v/@planetarium/tx
[Libplanet]: https://libplanet.io/
