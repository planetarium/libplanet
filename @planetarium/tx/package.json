{
  "name": "@planetarium/tx",
  "private": true,
  "description": "Creating Libplanet transactions from JavaScript/TypeScript",
  "main": "dist/pkg.cjs.js",
  "module": "dist/pkg.esm.js",
  "scripts": {
    "prebuild": "npm install --include=dev",
    "build": "bunchee --target es2020 --dts src/index.ts",
    "prepack": "npm run build",
    "dev": "bunchee --target es2020 --watch src/index.ts",
    "test": "dotnet build ../../Libplanet.Tools && vitest run",
    "coverage": "vitest run --coverage"
  },
  "types": "dist/src/index.d.ts",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/planetarium/libplanet.git",
    "directory": "@planet/tx"
  },
  "keywords": [
    "libplanet"
  ],
  "author": "Planetarium (https://planetarium.dev/)",
  "license": "LGPL-2.1-or-later",
  "bugs": {
    "url": "https://github.com/planetarium/libplanet/labels/js"
  },
  "homepage": "https://github.com/planetarium/libplanet/tree/main/@planet/tx",
  "devDependencies": {
    "@typescript-eslint/eslint-plugin": "^5.36.2",
    "@typescript-eslint/parser": "^5.36.2",
    "@vitest/coverage-c8": "^0.22.1",
    "@vitest/ui": "^0.22.1",
    "bunchee": "^2.1.3",
    "eslint": "^8.23.0",
    "eslint-config-prettier": "^8.5.0",
    "execa": "^6.1.0",
    "fast-check": "^3.1.2",
    "prettier": "2.7.1",
    "typescript": "^4.8.4",
    "vitest": "^0.22.1"
  },
  "dependencies": {
    "bencodex": "^0.1.1"
  }
}
