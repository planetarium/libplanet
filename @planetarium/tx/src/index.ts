export { type Address, encodeAddress } from "./address";
export {
  type Currency,
  encodeCurrency,
  type FungibleAssetValue,
  getCurrencyHash,
  getMajorUnit,
  getMinorUnit,
  getSign,
} from "./assets";
export {
  encodeMint,
  encodeSystemAction,
  encodeTransfer,
  type Mint,
  type SystemAction,
  type Transfer,
} from "./action";
export type { TxMetadata, UnsignedTxWithSystemAction } from "./tx";
export { encodeTxMetadata, encodeUnsignedTxWithSystemAction } from "./tx";
