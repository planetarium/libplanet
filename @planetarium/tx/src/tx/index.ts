export type { TxMetadata } from "./metadata";
export { encodeTxMetadata } from "./metadata";
export type { UnsignedTxWithSystemAction } from "./unsigned";
export { encodeUnsignedTxWithSystemAction } from "./unsigned";
