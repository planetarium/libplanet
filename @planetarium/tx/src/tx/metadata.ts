import { Encodable } from "bencodex";
import { Address, encodeAddress, encodeAddressSet } from "../address";
import { BlockHash, encodeBlockHash } from "../blockhash";
import { encodePublicKey, PublicKey } from "../key";

/**
 * Represents an unsigned transaction without actions.  Corresponds to
 * Libplanet's `TxMetadata`.
 */
export interface TxMetadata {
  nonce: bigint;
  publicKey: PublicKey;
  signer: Address;  // TODO: This field can be derived from publicKey.
  timestamp: Date;
  updatedAddresses: Set<Address>;
  genesisHash: BlockHash | null;
}

const NONCE_KEY = Buffer.from([0x6e]); // 'n'
const SIGNER_KEY = Buffer.from([0x73]); // 's'
const GENESIS_HASH_KEY = Buffer.from([0x67]); // 'g'
const UPDATED_ADDRESSES_KEY = Buffer.from([0x75]); // 'u'
const PUBLIC_KEY_KEY = Buffer.from([0x70]); // 'p'
const TIMESTAMP_KEY = Buffer.from([0x74]); // 't'

export function encodeTxMetadata(
  metadata: TxMetadata
): Map<string | Buffer,Encodable> {
  const updatedAddresses = encodeAddressSet(metadata.updatedAddresses);
  const timestamp = metadata.timestamp.toISOString()
    .replace(/Z$/, "000Z");
  const dict = new Map<string | Buffer, Encodable>([
    [NONCE_KEY, metadata.nonce],
    [SIGNER_KEY, encodeAddress(metadata.signer)],
    [UPDATED_ADDRESSES_KEY, updatedAddresses],
    [PUBLIC_KEY_KEY, encodePublicKey(metadata.publicKey)],
    [TIMESTAMP_KEY, timestamp],
  ]);
  if (metadata.genesisHash !== null) {
    dict.set(GENESIS_HASH_KEY, encodeBlockHash(metadata.genesisHash));
  }
  return dict;
}
