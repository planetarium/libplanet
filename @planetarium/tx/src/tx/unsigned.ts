import { Encodable } from "bencodex";
import { encodeSystemAction, SystemAction } from "../action";
import { encodeTxMetadata, TxMetadata } from "./metadata";

const SYSTEM_ACTION_KEY = Buffer.from([0x41]); // 'A'

export interface UnsignedTxWithSystemAction extends TxMetadata {
  systemAction: SystemAction;
}

/**
 * Encodes an unsigned transaction with a system action.
 * @param tx An unsigned transaction with a system action.
 * @returns An encoded transaction.
 */
export function encodeUnsignedTxWithSystemAction(
  metadata: UnsignedTxWithSystemAction 
): Encodable {
  const dict = encodeTxMetadata(metadata);
  dict.set(SYSTEM_ACTION_KEY, encodeSystemAction(metadata.systemAction));
  return dict;
}
