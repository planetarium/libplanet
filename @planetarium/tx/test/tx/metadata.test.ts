import { encode } from "bencodex";
import { expect, test } from "vitest";
import { encodeTxMetadata } from "../../src/tx/metadata";
import { key1 } from "./fixtures";

test("encodeTxMetadata", () => {
  const encoded = encodeTxMetadata({
    nonce: 123n,
    publicKey: key1.public,
    signer: key1.address,
    timestamp: new Date("2022-05-23T01:02:00+00:00"),
    updatedAddresses: new Set(),
    genesisHash: null,
  });
  expect(encoded).toMatchSnapshot();
  expect(encode(encoded)).toMatchSnapshot();
});
