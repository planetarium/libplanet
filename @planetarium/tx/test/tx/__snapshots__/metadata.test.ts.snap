// Vitest Snapshot v1

exports[`encodeTxMetadata 1`] = `
Map {
  {
    "data": [
      110,
    ],
    "type": "Buffer",
  } => 123n,
  {
    "data": [
      115,
    ],
    "type": "Buffer",
  } => ArrayBuffer [],
  {
    "data": [
      117,
    ],
    "type": "Buffer",
  } => [],
  {
    "data": [
      112,
    ],
    "type": "Buffer",
  } => ArrayBuffer [],
  {
    "data": [
      116,
    ],
    "type": "Buffer",
  } => "2022-05-23T01:02:00.000000Z",
}
`;

exports[`encodeTxMetadata 2`] = `
{
  "data": [
    100,
    49,
    58,
    110,
    105,
    49,
    50,
    51,
    101,
    49,
    58,
    112,
    51,
    51,
    58,
    2,
    0,
    224,
    39,
    9,
    204,
    12,
    5,
    29,
    193,
    5,
    24,
    140,
    69,
    74,
    46,
    126,
    247,
    179,
    107,
    133,
    218,
    52,
    82,
    157,
    58,
    188,
    25,
    104,
    22,
    124,
    245,
    79,
    49,
    58,
    115,
    50,
    48,
    58,
    38,
    131,
    68,
    186,
    70,
    230,
    202,
    42,
    138,
    80,
    150,
    86,
    85,
    72,
    185,
    1,
    139,
    198,
    135,
    206,
    49,
    58,
    116,
    117,
    50,
    55,
    58,
    50,
    48,
    50,
    50,
    45,
    48,
    53,
    45,
    50,
    51,
    84,
    48,
    49,
    58,
    48,
    50,
    58,
    48,
    48,
    46,
    48,
    48,
    48,
    48,
    48,
    48,
    90,
    49,
    58,
    117,
    108,
    101,
    101,
  ],
  "type": "Buffer",
}
`;
